andela/knights-bn-backend
# #170947662 Reset user password via email

#### What does this PR do?
 This PR ensures that the user can restore his/her password using email. 
#### Description of Task to be completed?
- once a user forgets his login credentials, he will be able to reset his password using the email he used to create his account in the system.
- test the endpoint
- Document feature using swagger

#### How should this be manually tested?
- run `git clone https://github.com/andela/knights-bn-backend.git` to clone the repo.
- run `cd knights-bn-backend` in your terminal
- run `git checkout ft-reset-password-170947661` to switch to this branch.
- run `npm install` to install dependencies.
- create a Postgres database.
- create a SendGrid account.
- create an API key under settings on   `API Keys`
```
- create an API key under settings on   `API Keys`
- click create API key then give it a name `BN_API_KEY`
- select Restricted access 
- scroll down where it is written Mail send then give it full access.
- scroll down then click on Create and View.
- then you will see an API key 
```
- create a .env file then add environment variable as shown below:
```
DB_USER='name of the user of the database'
DB_PASS='password of your Postgres'
DB_NAME='name of the database || Barefoot'
DB_TEST_NAME='name of the database for test || Barefoot_test'
NODE_ENV='development'
SECRETKEY='PrivateKey'
BN_API_KEY='use the SendGrid given ken'
```
- run `npm run createTables` to create the database tables.
- run `npm start` to start the server.
- open the browser use URL:`http://localhost:4000/api/v1/docs/#/User/Signup` for swagger documentation.
- Open postman and copy this URL POST `localhost:4000/api/v1/reset_pw/user`
- Add a comment body in the format shown below:
```
{
	"email": "the email of the password you want to recover",
}
```
- An email containing a link for resetting will be sent to the provided email, use the link before one hour. 
- Open postman PATCH `paste the link given in email`
- Add a comment body in the format shown below:
```
{
	"newPassword": "enter the new password",
         "confirmPassword": "confirm the new password",
}
```
- You Will see a message ` "message": "Password changed successfully" ` when the password is successfully changed.
#### Any background context you want to provide?
N/A

#### What are the relevant pivotal tracker stories?
[#170947661](https://www.pivotaltracker.com/story/show/170947661)

#### Screenshots

  ##Resetting password
 
      Password reset link sent successfully 

<img width="1141" alt="email send success" src="https://user-images.githubusercontent.com/50768983/74454628-d8541b00-4e8c-11ea-939c-8ae165c913ad.png">

      Strong password required on the new password creation.

<img width="1138" alt="invalid password 422" src="https://user-images.githubusercontent.com/50768983/74455027-6cbe7d80-4e8d-11ea-8ab0-a15b51e0771d.png">

      Confirmation password must match with the new password

<img width="1141" alt="confirmation not match" src="https://user-images.githubusercontent.com/50768983/74455126-8cee3c80-4e8d-11ea-8cb8-755175dba0f5.png">


      Password successfully changed.

<img width="1142" alt="password  success" src="https://user-images.githubusercontent.com/50768983/74455184-9e374900-4e8d-11ea-969b-0d245759d032.png">
